source/language/openpulse.rst

```diff
@@ -0,0 +1,317 @@
+.. code-block: javascript
+
+   {
+     drive: {
+       "{0}": "channel0",
+       "{1}": "channel1"
+     coupling: {
+       "{0,1}": "channel2"
+     }
+   }
+
+This has the advantage that one can run any program with any arbitrary frame
+names provided a mapping to the appropriate channels is given.
+
+Restrictions on the use of frames
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are two important considersations when dealing with frames.
+
+First, frames are compile time constants. Runtime variables cannot be used as
+arguments to frames; the frame must be resolvable at compile time. This also
+means that assigning a frame to an alias (for the purpose of avoiding typing)
+needs to be done with ``const`` not ``let``.
+
+.. code-block: none
+
+   const driveframe = frameof("drive", %0);
+
+Second, frames return references not values. This means that calling the same
+``frameof`` command in different places is the same as assigning the frame to a
+constant and referencing it multiple times.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase, scale) of a frame can be manipulated throughout program
+by referencing ``.frequency``, ``.phase``, and ``.scale``. Operations must be 
+appropriate for the respective type, ``float`` for frequency, ``angle`` for
+phase, and ``float`` for scale. Again, the exact precision of these calculations
+is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate:
+
+.. code-block:: none
+
+   // Shift phase of qubit 0's "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   frameof("drive", %0).phase += pi/4;
+
+   // Equivalent
+   const drive = frameof("drive", %0);
+   drive.phase += pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+   defcal rz(angle[20]:theta) %q {
+     frameof("drive", %q).phase -= theta;
+   }
+
+Here's an example qubit spectroscopy experiment.
+
+.. code-block:: none
+
+   qubit q;
+
+   const shots = 1000;
+   const start = 5e9; // 5 GHz
+   const end = 6e9; // 6 GHz
+   const points = 50;
+
+   complex[32] iq, average;
+   complex[32] output[points];
+
+   const driveframe = frameof("drive", q);
+
+   for p in [0 : points-1] {
+     // The key line: pick the frequency to sample
+     driveframe.frequency = start + (end-start) * p / points;
+     output[p] = 0;
+
+     for i in [0 : shots-1] {
+       // Assumes suitable calibrations for reset, x, and measure_iq
+       // and that the x gate references the same "drive" frame
+       reset q;
+       x q;
+       measure_iq q -> iq;
+
+       output[p] = (output[p] * i + iq) / (i + 1);
+     }
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: none
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Pulses
+------
+
+Pulses have two representations:
+
+- An array of complex samples which define the points for the pulse envelope
```

Comment:
It might be clearer to have a single `pulse` type, which is either an array of samples or the abstract mathematical function?